proposed/2022/vulnerabilities-in-restore.md

```diff
@@ -0,0 +1,406 @@
+  </suppressingVulnerabilities>
+```xml
+
+An idea how the project side could look like:
+Today, you can suppress warnings by using their warning codes at both the project and the package level.
+
+Example:
+
+```xml
+  <PropertyGroup>
+    <NoWarn>$(NoWarn);NU1901</NoWarn>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageId" Version="3.0.0" NoWarn="NU1901" />
+  </ItemGroup>
+
+```
+
+We can extend a similar functionality to the advisory urls via their own dedicated property.
+
+```xml
+<RestoreSuppressedAdvisoryUrls>https://github.com/advisories/GHSA-g3q9-xf95-8hp5</RestoreSuppressedAdvisoryUrls>
+
+ <PropertyGroup>
+    <RestoreSuppressedAdvisoryUrls>$(NoWarn);https://github.com/advisories/GHSA-g3q9-xf95-8hp5</RestoreSuppressedAdvisoryUrls>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="PackageId" Version="3.0.0" SuppressedAdvisoryUrls="https://github.com/advisories/GHSA-g3q9-xf95-8hp5" />
+  </ItemGroup>
+```
+
+The project level and package level metadata are not necessarily a package deal, we can choose to only add one.
+
+### Allow sources to upstream vulnerability information
+
+While the NuGet tooling does support multiple sources and a means to control which packages come from which source through [Package Source Mapping](https://learn.microsoft.com/en-us/nuget/consume-packages/package-source-mapping), all configurations are not going to contain nuget.org, and thus won't easily get the data curated from the [GitHub Advisory Database](https://devblogs.microsoft.com/nuget/how-to-scan-nuget-packages-for-security-vulnerabilities/).
+
+The [V3 protocol](https://learn.microsoft.com/en-us/nuget/api/overview) and the resource architecture would allow a source to link to a resource that may not be hosted by them.
+Not every source would want to setup a pipeline for processing vulnerability information. They can instead just "point" to the nuget.org information.
+
+For example, an Azure Artifacts source could have this in their index.json:
+
+```json
+{
+  "resources" [
+
+  {
+    "@id": "https://api.nuget.org/v3/vulnerability.json",
+    "@type": "VulnerabilityInfo/6.3.0",
+    "comment": "The base data for vulnerability update periodically"
+  }
+  ]
+}
+```
+
+While the nuget.org data only contains vulnerability information about packages on nuget.org, often times an `upstream` source is setup by customers, and while they reference their own private feed, they can still download packages from nuget.org. The vulnerability information is valuable in these scenarios.
+
+If this upstream approach is taken, it is very likely that the vulnerability checking has a lot of misses, but this is an understood trade-off.
+
+It is **very important* that if the upstream sources link to the nuget.org vulnerability, the actual nuget.org urls are used wherever possible, as this would allow the client to easily deduplicate the vulnerability information.
+
+### Command Line Restore output

```

Comment:
@JonDouglas FYI, I moved the vulnerability summary output on command line restores under "Future Possibilities" for two reasons. 1. So we can get a V1 of this feature out sooner. 2. I don't believe this proposal has been specced out well. I have a bunch of questions if we want to implement it.

For example, in the sample output, if there are non-vulerability warnings or errors, where in the output do they go?  Before the vulnerability summary, or below with the sample vulerability warnings?   If before, should vulerability warnings/errors be output twice, once between the summary once after?

What's the expected order of output when restoring a solution with many projects? Wait until all projects finish restoring, and then output all their summaries?  Or output project summaries as each project finishes restoring, even if it means that the vulnerability summary for that project will be somewhere much earlier in the restore log?

When restoring a solution where the same package version is in multiple projects, should we de-duplicate the output?

Anyway, as I said, I think we can take this on in the future if we still think it's useful. Let's get a V1 of the feature done without it, and see how we and customers feel about it.

Reply:
I think some details, like the ones about the handling of multiple vulnerability for a package can still be in the main part as that's likely going to be done.

Reply:
Let's see what this shapes to be while working to answer the unknown. Thanks for being so thorough with questions to be answered.